src/internal/m365/graph/service.go

```diff
@@ -275,3 +288,70 @@ func kiotaMiddlewares(
 
 	return mw
 }
+
+// ---------------------------------------------------------------------------
+// Graph Api Adapter Wrapper
+// ---------------------------------------------------------------------------
+
+var _ abstractions.RequestAdapter = &adapterWrap{}
+
+// adapterWrap takes a GraphRequestAdapter and replaces the Send() function to
+// act as a middleware for all http calls.  Certain error conditions never reach
+// the the client middleware layer, and therefore miss out on logging and retries.
+// By hijacking the Send() call, we can ensure three basic needs:
+// 1. Panics generated by the graph client are caught instead of crashing corso.
+// 2. Http and Http2 connection closures are retried.
+// 3. Error and debug conditions are logged.
+type adapterWrap struct {
+	abstractions.RequestAdapter
+}
+
+func wrapAdapter(gra *msgraphsdkgo.GraphRequestAdapter) *adapterWrap {
+	return &adapterWrap{gra}
+}
+
+var connectionEnded = filters.Contains([]string{
+	"connection reset by peer",
+	"client connection force closed",
+})
+
+func (aw *adapterWrap) Send(
+	ctx context.Context,
+	requestInfo *abstractions.RequestInformation,
+	constructor serialization.ParsableFactory,
+	errorMappings abstractions.ErrorMappings,
+) (sp serialization.Parsable, err error) {
+	defer func() {
+		if crErr := crash.Recovery(ctx, recover(), "graph adapter request"); crErr != nil {
+			err = Stack(ctx, crErr)
+		}
+	}()
+
+	// stream errors from http/2 will fail before we reach
+	// client middleware handling, therefore we don't get to
+	// make use of the retry middleware.  This external
+	// retry wrapper is unsophisticated, but should only
+	// retry in the event of a `stream error`, which is not
+	// a common expectation.
+	for i := 0; i < 3; i++ {
+		ictx := clues.Add(ctx, "request_retry_iter", i)
+
+		sp, err = aw.RequestAdapter.Send(ctx, requestInfo, constructor, errorMappings)
+		if err != nil &&
+			!(errors.Is(err, syscall.ECONNRESET) ||
```

Comment:
We do have a helper `IsErrConnectionReset` for this

Reply:
Yup, but I'm not currently certain that I want to catch context cancellations at this location, since those _should_ get handled by the middleware.  In the meantime I wanted to avoid 9x retries per call if we pathologically hit a ctx cancel.